lib/platform-stats.js

```diff
@@ -0,0 +1,18 @@
+import createDebug from 'debug'
+
+const debug = createDebug('spark:platform-stats')
+
+/**
+ * @param {import('pg').Client} pgClient
+ * @param {import('./preprocess').Measurement[]} honestMeasurements
+ */
+export const updateDailyNodeMetrics = async (pgClient, honestMeasurements) => {
+  debug('Updating daily node metrics, count=%s', honestMeasurements.length)
+  for (const m of honestMeasurements) {
+    await pgClient.query(`
+      INSERT INTO daily_node_metrics (station_id, metric_date)
+      VALUES ($1, now()::date)
+      ON CONFLICT (station_id, metric_date) DO NOTHING
+    `, [m.station_id]) // TODO: when we add more fields, we should update the ON CONFLICT clause to update the fields
```

Comment:
I have a concern about the storage required by this DB schema design. We will store each station ID string in many duplicates (one copy each day).

When I implemented "daily participants," I decided to create a lookup table mapping participant addresses to 32-bit integers and then store those integers in the `daily_participants` table using a foreign-key constraint.

See https://github.com/filecoin-station/spark-evaluate/pull/133

The key insight is that for most queries, we don't care what the exact participant address (or station ID) is. We want to count the number of unique addresses/IDs, and that can be achieved by counting the unique FK references.

The downside is that maintaining such a lookup table efficiently requires complex code (see `mapParticipantsToIds()` in my PR). OTOH, I think it should be fairly easy to generalise that function to support both participant addresses and station IDs by accepting a few more parameters like table name.

WDYT?

Reply:
I think this makes sense! As devil's advocate though with some napkin math:

If we're averaging about 2000 station nodes per day, and each row is, let's say, 250B, that's 174MB per year in total. A lookup table for station_id (and maybe deployment_type, CPU architecture, etc.) would save us roughly 60-100 bytes per row? So we'd save 55MB per year.

Even if we 15x the number of daily nodes, we're still looking at <1GB per year of data accumulation in savings. Is it worth the relatively complex code logic?

It's possible I missed some detail here or in the math though.

Reply:
Yeah, I was thinking, too, whether this optimisation is worth the complexity.

We already have 12-20k Station nodes. I think it's realistic to grow the network to 100k nodes by the end of this year. 

IIUC, Ed25519 has 32-byte long public keys, serialised as hex in 64 characters. We need 4 bytes to store the foreign key reference. 64-60=4. If we save 60 bytes per node, that gives us 365 * 60 * 100k bytes = ~2.19GB per year.

I think we can afford to store an extra 2 GB per module per year in our DB.

In the past, when we almost ran out of disk space, it was because our DB needed more than 500 GB. That's a long way to go :)

In that light, I propose to keep your current design.

@juliangruber WDYT?

Reply:
It should be a constant 64, no? Any reason to accept anything below or above that?

Reply:
Agreed that the optimization isn't worth it at this point 👍 And we should be able to easily migrate later

> It should be a constant 64, no? Any reason to accept anything below or above that?

+1 to making it the validation as strict as possible

Reply:
IMO, we should implement the same validation in `zinniad`, to reject invalid Station ID right during the startup. We should also change the hardcoded station id `zinnia` to use a valid string. It doesn't have to be a real public key, but it should pass the validation in spark-evaluate. For example, 64 zero characters (`000(...)000`) or something similar to Ethereum's null address `0x000(...)00dead`. The goal is to use an address that's easy to spot and filter out from production data.

Reply:
For (1), sure I can definitely work on this in Zinnia!

For (2), this was already done here: https://github.com/filecoin-station/spark-evaluate/blob/f86658b74e921f1b0778194d474dc10d352113de/lib/preprocess.js#L204-L210

Or am I misunderstanding you?

Reply:
> For (2), this was already done

Oh, I missed that bit. All is good! 👍🏻 

> For (1), sure I can definitely work on this in Zinnia!

Great! 💪🏻